rosutil/symdebug.rkt

```diff
@@ -0,0 +1,58 @@
+#lang racket
+
+(require (prefix-in r: rosette/safe))
+
+; Utilities for inspecting symbolic expressions in Yosys module state
+
+(provide term-length analyze-terms wide-write)
+
+; helper function that returns flattened length of an arbitrarily nested list
+(define (flattened-length l)
+  (if (not (list? l))
+      1
+      (if (empty? l)
+          0
+          (+ (flattened-length (first l))
+             (flattened-length (rest l))))))
+
+; returns "length" of (the possibly symbolic) t, depending on type of t:
+; vector? returns sum of term-length of each element
+; [symbolic] expression? returns flattened-length of t
+; [symbolic] constant? returns 1
+; other (i.e., concrete value) returns 0
+(define (term-length t)
+  (if (vector? t)
+      (foldl + 0 (map term-length (vector->list t)))
+      (if (r:expression? t)
+          (flattened-length (r:term->datum t))
+          (if (r:term? t)
+              1
+              0))))
+
+; analyze each element in state that is accessed via state-getters
+; state-getters is same construction as the argument of the same name to
+; verify-deterministic-start in rtl/shiva
+;
+; returns number of terms with length >= threshold
+; additionally outputs information about long state components if output? is #t
+(define (analyze-terms state state-getters
+                       #:threshold [threshold 3]
+                       #:output? [output? #f])
+  (define term-lengths
+    (for/list ([state-getter state-getters])
+      (let* ([name (first state-getter)]
+             [getter (second state-getter)]
+             [term (getter state)]
+             [len (term-length term)]
+             [symvars (r:symbolics term)])
+        (when (and output? (>= len threshold))
+          (printf "~a: expression size: ~a symbolics: ~a ~n" name len symvars))
+        len)))
+  (define big-term-count
+    (length (filter (lambda (n) (>= n threshold)) term-lengths)))
+  big-term-count)
+
+; display long symbolic expressions without truncating
+(define (wide-write x [width 1000])
+  (parameterize ([error-print-width width])
```

Comment:
Fair enough. I included it because I kept forgetting this exact one line and found myself digging it up in my Git history a few times, although I think I've internalized it a bit better at this point. I also didn't realize that this parameterization works with other display calls - just played around with it some more. so that's good to know!